pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+			ensure!(T::Pool::pool_exists(pool_id), Error::<T>::PoolNotFound);
+			Ok(())
+		}
+
+		fn generate_loan_id() -> T::LoanId {
+			LastLoanId::<T>::mutate(|last_loan_id| {
+				*last_loan_id = T::Hasher::hash(&*last_loan_id.as_ref());
+				*last_loan_id
+			})
+		}
+
+		fn to_rate_per_sec(rate_per_year: T::Rate) -> Result<T::Rate, DispatchError> {
+			T::InterestAccrual::convert_additive_rate_to_per_sec(rate_per_year)
+		}
+
+		fn find_write_off_state(
+			pool_id: PoolIdOf<T>,
+			maturity_date: Moment,
+		) -> Result<WriteOffState<T::Rate>, DispatchError> {
+			WriteOffState::find_best(
+				WriteOffPolicy::<T>::get(pool_id).into_iter(),
+				maturity_date,
+				T::Time::now().as_secs(),
+			)
+			.ok_or(Error::<T>::NoValidWriteOffState.into())
+		}
+
+		fn update_portfolio_valuation_with_pv(
+			pool_id: PoolIdOf<T>,
+			portfolio: &mut PortfolioValuation<T::Balance>,
+			old_pv: T::Balance,
+			new_pv: T::Balance,
+		) -> DispatchResult {
+			let prev_value = portfolio.value();
+
+			portfolio.update_with_pv_diff(old_pv, new_pv)?;
+
+			if prev_value != portfolio.value() {
+				Self::deposit_event(Event::<T>::PortfolioValuationUpdated {
+					pool_id,
+					value: portfolio.value(),
+					update_type: PortfolioValuationUpdateType::Inexact,
+				});
+			}
+
+			Ok(())
+		}
+
+		fn portfolio_valuation_for_pool(pool_id: PoolIdOf<T>) -> Result<T::Balance, DispatchError> {
+			ActiveLoans::<T>::get(pool_id).into_iter().try_fold(
+				T::Balance::zero(),
+				|sum, loan| -> Result<T::Balance, DispatchError> {
+					Ok(sum.ensure_add(loan.present_value()?)?)
+				},
+			)
+		}
+
+		fn make_active_loan<F, R>(
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			created_loan: CreatedLoan<T>,
+			f: F,
+		) -> Result<R, DispatchError>
+		where
+			F: FnOnce(&mut ActiveLoan<T>) -> Result<R, DispatchError>,
+		{
+			LatestPortfolioValuations::<T>::try_mutate(pool_id, |portfolio| {
+				ActiveLoans::<T>::try_mutate(pool_id, |active_loans| {
+					let mut loan = ActiveLoan::new(
+						loan_id,
+						created_loan.info,
+						created_loan.borrower,
+						Self::now(),
+					);
+
+					let result = f(&mut loan);
+					let new_pv = loan.present_value()?;
+					Self::update_portfolio_valuation_with_pv(
+						pool_id,
+						portfolio,
+						Zero::zero(),
+						new_pv,
+					)?;
+
+					active_loans
+						.try_push(loan)
+						.map_err(|_| Error::<T>::MaxActiveLoansReached)?;
+
+					result
+				})
+			})
+		}
+
+		fn update_active_loan<F, R>(
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			f: F,
+		) -> Result<R, DispatchError>
+		where
+			F: FnOnce(&mut ActiveLoan<T>) -> Result<R, DispatchError>,
+		{
+			LatestPortfolioValuations::<T>::try_mutate(pool_id, |portfolio| {
+				ActiveLoans::<T>::try_mutate(pool_id, |active_loans| {
+					let loan = active_loans
+						.iter_mut()
+						.find(|loan| loan.loan_id() == loan_id)
+						.ok_or(Error::<T>::LoanNotFound)?;
+
+					loan.update_time(portfolio.last_updated());
```

Comment:
Before trying to compute the `old_pv`, we adjust the `loan.last_updated` to the `portfolio.last_updated` in order to leave `portfolio_valuation_for_pool` inmutable. Although I think we will lose this property with #1193